agent/api/task.go

```diff
@@ -778,99 +791,237 @@ func (task *Task) SetKnownStatus(status TaskStatus) {
+// this field was updated or not
+func (task *Task) SetPullStartedAt(timestamp time.Time) bool {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	// Only set this field if it is not set
+	if task.PullStartedAt == (time.Time{}) {
+		task.PullStartedAt = timestamp
+		return true
+	}
+	return false
+}
+
+// GetPullStartedAt returns the PullStartedAt timestamp
+func (task *Task) GetPullStartedAt() time.Time {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.PullStartedAt
+}
+
+// SetPullStoppedAt sets the task pullstoppedat timestamp
+func (task *Task) SetPullStoppedAt(timestamp time.Time) {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	task.PullStoppedAt = timestamp
+}
+
+// GetPullStoppedAt returns the PullStoppedAt timestamp
+func (task *Task) GetPullStoppedAt() time.Time {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.PullStoppedAt
+}
+
+// SetExecutionStoppedAt sets the ExecutionStoppedAt timestamp of the task
+func (task *Task) SetExecutionStoppedAt(timestamp time.Time) {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	task.ExecutionStoppedAt = timestamp
+}
+
+// GetExecutionStoppedAt returns the task executionStoppedAt timestamp
+func (task *Task) GetExecutionStoppedAt() time.Time {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.ExecutionStoppedAt
+}
+
+// SetPullStartedAtSent marks the task PullStartedAt timestamp is sent
+func (task *Task) SetPullStartedAtSent() {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	task.PullStartedAtSent = true
+}
+
+// SetPullStoppedAtSent marks the task PullStoppedAt timestamp is sent
+func (task *Task) SetPullStoppedAtSent() {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	task.PullStoppedAtSent = true
+}
+
+// SetExecutionStoppedAtSent marks the task ExecutionStoppedAt timestamp is sent
+func (task *Task) SetExecutionStoppedAtSent() {
+	task.statusLock.Lock()
+	defer task.statusLock.Unlock()
+
+	task.ExecutionStoppedAtSent = true
+}
+
+// GetPullStartedAt returns whether the task PullStartedAt timestamps is sent
+func (task *Task) GetPullStartedAtSent() bool {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.PullStartedAtSent
+}
+
+// GetPullStoppedAt returns whether the task PullStoppedAt timestamps is sent
+func (task *Task) GetPullStoppedAtSent() bool {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.PullStoppedAtSent
+}
+
+// GetExecutionStoppedAt returns whether the task ExecutionStoppedAt timestamps is sent
+func (task *Task) GetExecutionStoppedAtSent() bool {
+	task.statusLock.RLock()
+	defer task.statusLock.RUnlock()
+
+	return task.ExecutionStoppedAtSent
+}
+
+// ShouldSendExecutionStoppedTime determins whether the container state change
+// should cause ExecutionStoppedAt to be sent
+func (task *Task) ShouldSendExecutionStoppedTime(container *Container, timestamp time.Time) bool {
```

Comment:
You'd need to change this based on my recommendation for getting rid of the timer for this.